build/source/engine/allocspace.f90

```diff
@@ -106,36 +106,71 @@ subroutine allocate_gru_struc(nGRU,nHRU,err,message)
-
- ! get hru_dim dimension length (ie., the number of hrus in entire domain)
- err = nf90_inq_dimid(ncid, "nhru", hruDimID);             if(err/=0)then; message=trim(message)//'problem finding nHRU dimension'; return; endif
- err = nf90_inquire_dimension(ncid, hruDimID, len = nHRU); if(err/=0)then; message=trim(message)//'problem reading nHRU dimension'; return; endif
  
- allocate(gru_struc(nGRU), index_map(nHRU), stat=err)
- if(err/=0)then; err=20; message=trim(message)//'problem allocating space for mapping structures'; return; endif
+ if (strtHRU<=0) then
+  ! this is a full run. 
+  ! get hru_dim dimension length (ie., the number of hrus in entire domain)
+  err = nf90_inq_dimid(ncid, "hru", hruDimID);             if(err/=0)then; message=trim(message)//'problem finding hru dimension'; return; endif
+  err = nf90_inquire_dimension(ncid, hruDimID, len = nHRU); if(err/=0)then; message=trim(message)//'problem reading hru dimension'; return; endif
  
- allocate(gru_id(nGRU),hru_id(nHRU),hru2gru_id(nHRU),stat=err)
+ end if 
+ ! read the hruID and HRU's gruID
+ allocate(gru_id(nHRU),hru_id(nHRU),hru2gru_id(nHRU),index_map(nHRU),stat=err)
  if(err/=0)then; err=20; message=trim(message)//'problem allocating space for zLocalAttributes gru-hru correspondence vectors'; return; endif
-
- ! read gru_id from netcdf file
- err = nf90_inq_varid(ncid, "gruId", varid);     if(err/=0)then; message=trim(message)//'problem finding gruId variable'; return; endif
- err = nf90_get_var(ncid,varid,gru_id);          if(err/=0)then; message=trim(message)//'problem reading gruId variable'; return; endif
-
- ! read the HRU to GRU mapping
+ 
+ err = nf90_inq_varid(ncid, "hruId", varid); if(err/=0)then; message=trim(message)//'problem finding hruId variable'; return; endif
+ err = nf90_get_var(ncid,varid,hru_id,start=(/max(strtHRU,1)/),count=(/nHRU/))
+ if(err/=0)then; message=trim(message)//'problem reading hruId variable'; return; endif  
+ 
  err = nf90_inq_varid(ncid, "hru2gruId", varid); if(err/=0)then; message=trim(message)//'problem finding hru2gruId variable'; return; endif
- err = nf90_get_var(ncid,varid,hru2gru_id);      if(err/=0)then; message=trim(message)//'problem reading hru2gruId variable'; return; endif
+ err = nf90_get_var(ncid,varid,hru2gru_id,start=(/max(strtHRU,1)/),count=(/nHRU/))
+ if(err/=0)then; message=trim(message)//'problem reading hru2gruId variable'; return; endif
+ 
+ ! extract unique gruID 
+ nGRU = 1
+ gru_id(1) = hru2gru_id(1) 
+ index_map(1)%gru_ix = nGRU
+ 
+ hruLoop: do iHRU=2,nHRU
+  do iGRU=1,nGRU
+   if (gru_id(iGRU) == hru2gru_id(iHRU)) then
+    index_map(iHRU)%gru_ix=iGRU
+    ! skip duplicate
+    cycle hruLoop
+   end if
+  end do
+  ! No match found so it is a unique GRU. add the GRU to the GRU list
+  nGRU = nGRU + 1
+  gru_id(nGRU) = hru2gru_id(iHRU)
+  index_map(iHRU)%gru_ix=nGRU
+ end do hruLoop
+ 
```

Comment:
I think that it would actually be better to parallelize at the GRU level instead of at the HRU level. I understand that this is non-trivial.